File: LtpaToken2.js
const CryptoJS = require("crypto-js");
const NodeRSA = require('node-rsa');

function eGcd (a, b) {
  a = BigInt(a)
  b = BigInt(b)
  if (a <= 0n | b <= 0n) throw new RangeError('a and b MUST be > 0') // a and b MUST be positive
  let x = 0n
  let y = 1n
  let u = 1n
  let v = 0n
  while (a !== 0n) {
    const q = b / a
    const r = b % a
    const m = x - (u * q)
    const n = y - (v * q)
    b = a
    a = r
    x = u
    y = v
    u = m
    v = n
  }
  return {
    g: b,
    x: x,
    y: y
  }
}

function toZn (a, n) {
  n = BigInt(n)
  if (n <= 0) { return NaN }
  a = BigInt(a) % n
  return (a < 0) ? a + n : a
}

function modInv (a, n) {
  try {
    const egcd = eGcd(toZn(a, n), n)
    if (egcd.g !== 1n) {
      return NaN // modular inverse does not exist
    } else {
      return toZn(egcd.x, n)
    }
  } catch (error) {
    return NaN
  }
}

function wordToByteArray(wordArray) {
    var byteArray = [], word, i, j;
    for (i = 0; i < wordArray.length; ++i) {
        word = wordArray[i];
        for (j = 3; j >= 0; --j) {
            byteArray.push((word >> 8 * j) & 0xFF);
        }
    }
    return byteArray;
}

function getSecretKey(key, password){
	const hash = CryptoJS.lib.WordArray.create(CryptoJS.SHA1(CryptoJS.enc.Utf8.parse(password)).words.concat(0));
	return CryptoJS.TripleDES.decrypt({ciphertext:CryptoJS.enc.Base64.parse(key)},hash,{mode:CryptoJS.mode.ECB});
}

function privKeyToRSA(key){
	//get available RSA components from private key
	const eLength = 3;
	const pLength = 65;
	const qLength = 65;
	const components = Array(8).fill(null);

	components[2] = Buffer.alloc(eLength);
	components[3] = Buffer.alloc(pLength);
	components[4] = Buffer.alloc(qLength);
	
	if (key.length > eLength + pLength + qLength) {
		const abyte2 = Buffer.alloc(4);
		key.copy(abyte2, 0, 0, 4);
		const privExponentLength = abyte2.readUIntBE(0, 4);
		components[1] = Buffer.alloc(privExponentLength);
		key.copy(components[1], 0, 4, privExponentLength + 4);
		key.copy(components[2], 0, privExponentLength + 4, privExponentLength + 4 + eLength);
		key.copy(components[3], 0, privExponentLength + 4 + eLength, privExponentLength + 4 + eLength + pLength);
		key.copy(components[4], 0, privExponentLength + 4 + eLength + pLength, privExponentLength + 4 + eLength + pLength + qLength);
	}
	else {
		key.copy(components[2], 0, 0, eLength);
		key.copy(components[3], 0, eLength, eLength + pLength);
		key.copy(components[4], 0, eLength + pLength , eLength + pLength + qLength);  
	}

	//compute missing RSA components
	const bigints = Array(8).fill(null);
	for (i = 0; i <= 7; ++i) {
		if (components[i]!=null)
			bigints[i]= BigInt('0x'+components[i].toString('hex'))
	}
	
	if (bigints[3]<bigints[4]) {
		let bigint = bigints[3];
		bigints[3] = bigints[4];
		bigints[4] = bigint;
		bigint = bigints[5];
		bigints[5] = bigints[6];
		bigints[6] = bigint;
		bigints[7] = null;
	}
	
	if (bigints[7] == null)
		bigints[7] = modInv(bigints[4],bigints[3]);
	if (bigints[0] == null)
		bigints[0] = bigints[3]*(bigints[4]);
	if (bigints[1] == null)
		bigints[1] = modInv(bigints[2],((bigints[3]-1n)*(bigints[4]-1n)));
	if (bigints[5] == null)
		bigints[5] = bigints[1]%(bigints[3]-1n);
	if (bigints[6] == null)
		bigints[6] = bigints[1]%(bigints[4]-1n);

	//construct the RSA object
	//n:		modulus				bigints[0]
	//d:		private exponent	bigints[1]
	//e:		public exponent		bigints[2]	
	//p:		prime1 				bigints[3]
	//q:		prime2				bigints[4]
	//dmp1: 	exponent1			bigints[5]
	//dmq1: 	exponent2			bigints[6]
	//coeff:	coefficient			bigints[7]
	const rsa = NodeRSA({b:1024});
	rsa.setOptions({environment:'node',signingScheme:'sha1'});
	rsa.importKey({n:bigints[0].toString(),d:bigints[1].toString(),e:Number(bigints[2].toString()),p:bigints[3].toString(),q:bigints[4].toString(),dmp1:bigints[5].toString(),dmq1:bigints[6].toString(),coeff:bigints[7].toString()},'components',);
	return rsa;
}

function pubKeyToRSA (key) {
	//get available RSA components from public key
	let buff = Buffer.from(key,'base64');
	const mod = Buffer.alloc(129);
	const exp = Buffer.alloc(3);
	buff.copy(mod, 0, 0,129);
	buff.copy(exp, 0, 129, 132);
	
	//construct the RSA object
	const rsa = NodeRSA({b:1024});
	rsa.setOptions({environment:'node',signingScheme:'sha1'});	
	rsa.importKey({n:Buffer.from(mod),e:exp,},'components-public');
	return rsa;
};

function signLtpaToken2 (data, rsa) {
	const hashedBody = Buffer.from(wordToByteArray(CryptoJS.SHA1(data).words));
	const signature = rsa.sign(hashedBody,'base64')
	return signature;		
};

function encryptLtpaToken2 (data,key) {
	key = CryptoJS.lib.WordArray.create(key.words.slice(0,4));
	return CryptoJS.enc.Base64.stringify(CryptoJS.AES.encrypt(data,key,{iv:key}).ciphertext);
}

function decryptLtpaToken2 (data,key) {
	key = CryptoJS.lib.WordArray.create(key.words.slice(0,4));
	return CryptoJS.AES.decrypt({ciphertext:CryptoJS.enc.Base64.parse(data)},key,{iv:key}).toString(CryptoJS.enc.Utf8);
}

function verifyLtpaToken2 (data, rsa) {
	const body = data[0];
	const expire = data[1];
	const signature = data[2];
	var retValue=false;
	const hashedBody = Buffer.from(wordToByteArray(CryptoJS.SHA1(body).words));
	const verification = rsa.verify(hashedBody,signature,'buffer','base64');
	const expiration = new Date()<new Date(expire*1);
	if (expiration&&verification)
		retValue=true;
	return retValue;
};

const _3DESKey="QeVYdNvQbz7jyqbFu3wmeuyws96KwmvBEu3o6+o138E\=";
const _PrivateKey="uHUSg2YvtKovgtQLX+SmtH4BPnyBy7cLnNsI+0QaC+KcMVKNuBYjYknyP0n+CCJgkDebdjz5vHqhqlg3abv/P19dzjvJCCHXzIDapYOPBBYcmWZGpMB19b6bsykwjdNbf+xjijRQvOXetf5///ljiHeq/NP58qpS9KXfyXcjXGdEAwFSKAFTG1bj9Cpy6iqWQ9SPFD3kiEhzNu16lSmR4BNtZTpZ0uy8hfYB1u9HB3/sJ0ih2iw7qR8fnhVuKbpIyAtio5sPOHfgayI01vDhEdHNPcZaTxx5Ndf1MXq05Bv2ZEX3JRMtVsLfOvNBnz5PdmPj74CH8Qy7oa4ZX2bDEWF9pBkS7B9rPKDe291/d7M\=";
const _PublicKey = "ALTw+Sy9dQSv8lQ6JPX/zhqwLtua6yo9mmrC55NAxu7SLXx2Ee+A8OBMTH4+4OIk0pnNAqfR8AKARY4D3fqEJB5z+V/6Zh9Gap3tGT7wmTf0mrtF9EqgLCiVqfBq+0LM+ZfvT6YC6PG1CFVM1kkuuvn2Sc2T+tuiTQSX+zWauR45AQAB";
const password="lotus123";

//decrypt encrypted 3DES key
const TrippleDESKey = getSecretKey(_3DESKey,password)

//decrypt encrypted private key
const PrivateKey = getSecretKey(_PrivateKey,password)

//get RSA object from private key
const rsa1 = privKeyToRSA((Buffer.from(wordToByteArray(PrivateKey.words))));

//calculate the epoch expiration, set at 2hours
var expire = Math.floor((new Date() / 1000)+7200) + "000";

//add expire to the body and add some fields like user DN etc
const body = "expire:"+expire+"$u:user\\:defaultWIMFileBasedRealm/CN=Pieter Geerts,O=beibm";

//calculate signature for LtpaToken2 as BASE64(SHA1_WITH_RSA(SHA_DIGEST(body)))
const signature = signLtpaToken2(body,rsa1);

//raw token as body%expiration%signature
const rawToken = [body, expire, signature].join('%');

//encrypt the raw token and return as base64 string
const encToken = encryptLtpaToken2(rawToken,TrippleDESKey);
console.log(encToken);

//decrypt the LtpaToken2
const rawToken2 = decryptLtpaToken2(encToken,TrippleDESKey);

//get RSA object from unencrypted public key
const rsa2 = pubKeyToRSA(_PublicKey);

//verify the LtpaToken and return the raw token
if (verifyLtpaToken2(rawToken2.split('%'),rsa2))
	console.log(rawToken2);

File: LtpaToken2.java
import java.math.BigInteger;
import java.security.KeyFactory;
import java.security.MessageDigest;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.Signature;
import java.security.spec.RSAPrivateCrtKeySpec;
import java.security.spec.RSAPublicKeySpec;
import java.util.Arrays;
import java.util.Base64;
import java.util.Date;

import javax.crypto.Cipher;

public class LtpaToken2 {

	private static byte[] getSecretKey(String key, String password) throws Exception {
		MessageDigest md = MessageDigest.getInstance("SHA");
		md.update(password.getBytes());
		byte[] hash3DES = new byte[24];
		System.arraycopy(md.digest(), 0, hash3DES, 0, 20);
		Arrays.fill(hash3DES, 20, 24, (byte) 0);
		Cipher cipher = javax.crypto.Cipher.getInstance("DESede/ECB/PKCS5Padding");
		cipher.init(javax.crypto.Cipher.DECRYPT_MODE, javax.crypto.SecretKeyFactory.getInstance("DESede").generateSecret(new javax.crypto.spec.DESedeKeySpec(hash3DES)));
		return cipher.doFinal(Base64.getDecoder().decode(key.getBytes()));
	}

	private static RSAPrivateCrtKeySpec privKeyToRSA(byte[] key) {
		// get available RSA components from private key
		int eLength = 3;
		int pLength = 65;
		int qLength = 65;
		byte[][] components = new byte[8][];

		components[2] = new byte[eLength];
		components[3] = new byte[pLength];
		components[4] = new byte[qLength];

		if (key.length > eLength + pLength + qLength) {
			byte[] abyte2 = new byte[4];
			for (int i = 0; i < 4; i++) {
				abyte2[i] = key[i];
			}
			int privExponentLength = ((abyte2[0] & 0xFF) << 24) | ((abyte2[1] & 0xFF) << 16) | ((abyte2[2] & 0xFF) << 8) | ((abyte2[3] & 0xFF) << 0);
			components[1] = new byte[privExponentLength];
			System.arraycopy(key, 4, components[1], 0, privExponentLength);
			System.arraycopy(key, privExponentLength + 4, components[2], 0, eLength);
			System.arraycopy(key, privExponentLength + 4 + eLength, components[3], 0, pLength);
			System.arraycopy(key, privExponentLength + 4 + eLength + pLength, components[4], 0, qLength);
		} else {
			System.arraycopy(key, 0, components[2], 0, eLength);
			System.arraycopy(key, eLength, components[3], 0, pLength);
			System.arraycopy(key, eLength + pLength, components[4], 0, qLength);
		}

		// compute missing RSA components
		BigInteger[] bigints = new BigInteger[8];
		for (int i = 0; i <= 7; i++)
			if (components[i] != null)
				bigints[i] = new BigInteger(1, components[i]);
		if (bigints[3].compareTo(bigints[4]) < 0) {
			BigInteger bigint = bigints[3];
			bigints[3] = bigints[4];
			bigints[4] = bigint;
			bigint = bigints[5];
			bigints[5] = bigints[6];
			bigints[6] = bigint;
			bigints[7] = null;
		}
		if (bigints[7] == null)
			bigints[7] = bigints[4].modInverse(bigints[3]);
		if (bigints[0] == null)
			bigints[0] = bigints[3].multiply(bigints[4]);
		if (bigints[1] == null)
			bigints[1] = bigints[2].modInverse(bigints[3].subtract(BigInteger.ONE).multiply(bigints[4].subtract(BigInteger.ONE)));
		if (bigints[5] == null)
			bigints[5] = bigints[1].remainder(bigints[3].subtract(BigInteger.ONE));
		if (bigints[6] == null)
			bigints[6] = bigints[1].remainder(bigints[4].subtract(BigInteger.ONE));

		//construct the RSA object
		//n:		modulus				bigints[0]
		//d:		private exponent	bigints[1]
		//e:		public exponent		bigints[2]	
		//p:		prime1 				bigints[3]
		//q:		prime2				bigints[4]
		//dmp1: 	exponent1			bigints[5]
		//dmq1: 	exponent2			bigints[6]
		//coeff:	coefficient			bigints[7]
		return new java.security.spec.RSAPrivateCrtKeySpec(bigints[0], bigints[2], bigints[1], bigints[3], bigints[4], bigints[5], bigints[6], bigints[7]);
	}

	private static RSAPublicKeySpec pubKeyToRSA(String key) {
		//get available RSA components from public key
		byte[] parts = Base64.getDecoder().decode(key);
		BigInteger modulus = new BigInteger(Arrays.copyOfRange(parts, 0, 129));
		BigInteger exponent = new BigInteger(Arrays.copyOfRange(parts, 129, 129 + 3));
		
		//construct the RSA object
		//RSAPublicKeySpec(BigInteger modulus, BigInteger publicExponent)
		RSAPublicKeySpec pubKeySpec = new java.security.spec.RSAPublicKeySpec(modulus, exponent);
		return pubKeySpec;
	}
	
	private static String signLtpaToken2(String data, RSAPrivateCrtKeySpec rsa) throws Exception {
		MessageDigest md1JCE = MessageDigest.getInstance("SHA");
		byte[] plainUserDataBytes = md1JCE.digest(data.getBytes());
		byte[] encodedSignatureBytes = null;
		PrivateKey privatekey = java.security.KeyFactory.getInstance("RSA").generatePrivate(rsa);
		Signature signer = java.security.Signature.getInstance("SHA1withRSA");
		signer.initSign(privatekey);
		signer.update(plainUserDataBytes, 0, plainUserDataBytes.length);
		encodedSignatureBytes = signer.sign();
		return Base64.getEncoder().encodeToString(encodedSignatureBytes);
	}

	private static String encryptLtpaToken2(String data, byte[] key) throws Exception {
		Cipher cipher = javax.crypto.Cipher.getInstance("AES/CBC/PKCS5Padding");
		cipher.init(javax.crypto.Cipher.ENCRYPT_MODE, new javax.crypto.spec.SecretKeySpec(key, 0, 16, "AES"), new javax.crypto.spec.IvParameterSpec(key, 0, 16));
		return Base64.getEncoder().encodeToString(cipher.doFinal(data.getBytes("UTF8")));
	}

	private static String decryptLtpaToken2(String data, byte[] key) throws Exception {
		Cipher cipher = javax.crypto.Cipher.getInstance("AES/CBC/PKCS5Padding");
		cipher.init(javax.crypto.Cipher.DECRYPT_MODE, new javax.crypto.spec.SecretKeySpec(key, 0, 16, "AES"), new javax.crypto.spec.IvParameterSpec(key, 0, 16));
		return new String(cipher.doFinal(Base64.getDecoder().decode(data)));
	}

	private static boolean verifyLtpaToken2(String[] data, RSAPublicKeySpec rsa) throws Exception {
		String body = data[0];
		String expire = data[1];
		String signature = data[2];
		boolean retValue = false;
		KeyFactory kf = java.security.KeyFactory.getInstance("RSA");
		PublicKey pubkey = kf.generatePublic(rsa);
		Signature signer = java.security.Signature.getInstance("SHA1withRSA");
		signer.initVerify(pubkey);
		byte[] hashedBody = java.security.MessageDigest.getInstance("SHA").digest(body.getBytes());
		signer.update(hashedBody);
		boolean verification = signer.verify(Base64.getDecoder().decode(signature));
		boolean expiration = (new Date(System.currentTimeMillis())).compareTo(new Date(Long.parseLong(expire))) < 0 ? true : false;
		if (expiration && verification)
			retValue = true;
		return retValue;
	}

	public static void main(String[] args) throws Exception {

		String _3DESKey = "QeVYdNvQbz7jyqbFu3wmeuyws96KwmvBEu3o6+o138E";
		String _PrivateKey = "uHUSg2YvtKovgtQLX+SmtH4BPnyBy7cLnNsI+0QaC+KcMVKNuBYjYknyP0n+CCJgkDebdjz5vHqhqlg3abv/P19dzjvJCCHXzIDapYOPBBYcmWZGpMB19b6bsykwjdNbf+xjijRQvOXetf5///ljiHeq/NP58qpS9KXfyXcjXGdEAwFSKAFTG1bj9Cpy6iqWQ9SPFD3kiEhzNu16lSmR4BNtZTpZ0uy8hfYB1u9HB3/sJ0ih2iw7qR8fnhVuKbpIyAtio5sPOHfgayI01vDhEdHNPcZaTxx5Ndf1MXq05Bv2ZEX3JRMtVsLfOvNBnz5PdmPj74CH8Qy7oa4ZX2bDEWF9pBkS7B9rPKDe291/d7M";
		String _PublicKey = "ALTw+Sy9dQSv8lQ6JPX/zhqwLtua6yo9mmrC55NAxu7SLXx2Ee+A8OBMTH4+4OIk0pnNAqfR8AKARY4D3fqEJB5z+V/6Zh9Gap3tGT7wmTf0mrtF9EqgLCiVqfBq+0LM+ZfvT6YC6PG1CFVM1kkuuvn2Sc2T+tuiTQSX+zWauR45AQAB";
		String password = "lotus123";

		// decrypt encrypted 3DES key
		byte[] TrippleDESKey = getSecretKey(_3DESKey, password);

		// decrypt encrypted private key
		byte[] PrivateKey = getSecretKey(_PrivateKey, password);

		// get RSA object from private key
		RSAPrivateCrtKeySpec rsa1 = privKeyToRSA(PrivateKey);

		// calculate the epoch expiration, set at 2hours
		String expire = ((int) (System.currentTimeMillis() / 1000) + 7200) + "000";

		// add expire to the body and add some fields like user DN etc
		String body = "expire:" + expire + "$u:user\\:defaultWIMFileBasedRealm/CN=Pieter Geerts,O=beibm";

		// calculate signature for LtpaToken2 as BASE64(SHA1_WITH_RSA(SHA_DIGEST(body)))
		String signature = signLtpaToken2(body, rsa1);

		// raw token as body%expiration%signature
		String rawToken = String.join("%", body, expire, signature);

		// encrypt the raw token and return as base64 string
		String encToken = encryptLtpaToken2(rawToken, TrippleDESKey);
		System.out.println(encToken);

		// decrypt the LtpaToken2
		String rawToken2 = decryptLtpaToken2(encToken, TrippleDESKey);

		// get RSA object from unencrypted public key
		RSAPublicKeySpec rsa2 = pubKeyToRSA(_PublicKey);

		// verify the LtpaToken and return the raw token
		if (verifyLtpaToken2(rawToken2.split("%"), rsa2)) {
			System.out.println(rawToken2);
		}
	}
}
